# Assignment 1

Link: https://github.com/AndreasRahbekPalm/CSS_assignments.git

Contributions: We have been collaburating equally on all of the parts.

by: Andreas Rahbek-Palm (s214598) and Valdemar Hee Hansen (s214621)

Code was made in the exercise week and have been made partly in collabaration with people who are in other groups

# Part 1: Web-scraping

## code:

In [25]:
# Import libraries
from bs4 import BeautifulSoup
import requests
import re
import pandas as pd
from tqdm import tqdm
from joblib import Parallel, delayed

In [26]:
# Get the HTML content of the page
link = "https://ic2s2-2023.org/program"
r = requests.get(link)
soup = BeautifulSoup(r.content)

In [12]:
table = soup.find("table",{"class":"tutorials"})
table_rows = table.find_all("tr")

# Here we find the keynote talks and store them in a list wow
keynotes = []
for tr in table_rows:
    td = tr.find_all("td")
    row = [i.text for i in td]
    if row:
        if "Keynote" in row[1]:
            keynotes.append(row[1])

print(keynotes)

# cleaning it up so we only have the names of the keynotes
keynotes = [re.sub("Keynote - ", "", i) for i in keynotes]

print(keynotes)


['Keynote - Jevin West', 'Keynote - Linda Steg', 'Keynote - Sharad Goel', 'Keynote - Molly Crockett', 'Keynote - Lisa Anne Hendriks', 'Keynote - Stefan Gössling', 'Keynote - Joanna Bryson', 'Keynote - Tim Althoff', 'Keynote - Lauren Brent', 'Keynote - Esteban Moro']
['Jevin West', 'Linda Steg', 'Sharad Goel', 'Molly Crockett', 'Lisa Anne Hendriks', 'Stefan Gössling', 'Joanna Bryson', 'Tim Althoff', 'Lauren Brent', 'Esteban Moro']


In [13]:
# we saw a pattern and get these so we get some chairs
pattern = r'Chair:\s+([A-Za-z\s]+?)(?=[,.\n])'

# Find all occurrences of the pattern
matches = re.findall(pattern, soup.get_text())

# Processing the matches to trim any trailing whitespace
chairs = [match.strip() for match in matches]

# Print the list or process as needed
print(chairs)

['Claudia Wagner', 'Taha Yasseri', 'David Garcia', 'Milena Tsvetkova', 'Elisa Omodei', 'Leo Ferres', 'Clara Vandeweerdt', 'Andreia Sofia Teixeira', 'Edward Lee', 'Luca Rossi', 'Jonas L Juul', 'Michele Coscia', 'Yelena Mejova', 'Michael Szell', 'Lisette Espin Noboa', 'Matteo Magnani', 'Onur Varol', 'Roberta Rocca', 'Talal Rahwan', 'Frauke Kreuter', 'Riccardo Di Clemente', 'Martin Ardvisson', 'Taha Yasseri', 'Michele Tizzoni', 'Daniele Quercia', 'Kokil Jaidka', 'Ingmar Weber', 'Claudia Acciai', 'Sandor Juhasz', 'Anna Sapienza', 'Yunkyu Sohn', 'Alexander Gates', 'Vedran Sekara', 'Piotr Sapiezynski', 'Daniela Paolotti', 'Pantelis Pipergias Analytis', 'Michele Starnini', 'Kristina Gligoric', 'Rossano Schifanella', 'Petter Holme', 'Kyriaki Kalimeri', 'Bruno Lepri', 'Ivan Smirnov', 'Yevgeniy Golovchenko', 'Clemens Stachl', 'Maximilian Schich', 'Kelton Minor']


In [14]:
names = []

# Find all <i> tags
italic_tags = soup.find_all('i')

# Extract names from each <i> tag because we saw it was in italics (but here we also have the chairs)
for tag in italic_tags:
    # Split the text by commas and strip whitespace
    names_in_tag = [name.strip() for name in tag.get_text().split(',')]
    names.extend(names_in_tag)

# Print the list or process as needed
print(names)

# Remove element in list if it starts with "Chair:" because otherwise we get a lot of duplicates
authors_posters = [name for name in names if not name.startswith('Chair:')]
print(authors_posters)

['Chair: Claudia Wagner', 'Jonas L Juul', 'Jon Kleinberg', 'Chloe Ahn', 'Xinyi Wang', 'Giuseppe Russo', 'luca verginer', 'Manoel Horta Ribeiro', 'Giona Casiraghi', 'Almog Simchon', 'Adam Sutton', 'Matthew Edwards', 'Stephan Lewandowsky', 'Arianna Pera', 'Manuel Vimercati', 'Matteo Palmonari', 'Mohammed Alsobay', 'Abdullah Almaatouq', 'David G. Rand', 'Duncan J. Watts', 'Sara Venturini', 'Satyaki Sikdar', 'Francesco Rinaldi', 'Francesco Tudisco', 'Santo Fortunato', 'Isabella Loaiza', 'Takahiro Yabe', 'Alex Pentland', 'Chair: Taha Yasseri', 'Silvia De Sojo Caso', 'Mia Ann Jørgensen', 'Sune Lehmann', 'Laura Alessandretti', 'Emil Bakkensen Johansen', 'Mathias Wullum Nielsen', 'Rubén Rodríguez Casañ', 'Antonio Ariño Villarroya', 'Sunny Rai', 'Ashley Francisco', 'Salvatore Giorgi', 'Brenda Curtis', 'Lyle Ungar', 'Sharath Chandra Guntuku', 'Allison Koenecke', 'Eric Giannella', 'Robb Willer', 'Sharad Goel', 'Ziv Epstein', 'Hause Lin', 'Levin Brinkmann', 'Bramantyo Supriyatno', 'Iyad Rahwan', '

In [15]:
print(len(keynotes))
print(len(chairs))  
print(len(authors_posters)) 

all = keynotes + chairs + authors_posters

print(len(all))
#find unique elements
all = set(all)
print(len(all))

df = pd.DataFrame(all, columns=['Name'])
df.to_csv('names.txt', index=False)

10
47
2037
2094
1490


#### 5 How many unique researchers do you get?

we get 1490 unique researchers

#### 6 Explain the process you followed to web-scrape the page. Which choices did you make to accurately retreive as many names as possible? Which strategies did you use to assess the quality of your final list? Explain your reasoning and your choices

We firstly isolated the first table where the keynotes where located. Then we identified a pattern for the once who chairs and isolated them. Lastly we saw that the all of the researchers where in italic text (and only the names and chairs where in italic) so we got all of the chairs and split by "," to isolate. lastly we use set() to get unique.  

# Part 2 Ready Made vs Custom Made Data

#### 1
cons centola: dont know the other so the connection is not strong between individuals (different behaviour than with friends). All of these individuals also has an intrest in health which makes it a more artificial enviroment as it is not representable for the entire population as diverce friendgroups / networks exists. 

pros centola: it is very controled as they control all of the information so they minimize unwanted causality and can focus only on the network structure.

Nicolaides's cons: some people might not have notifications on so the data becomes incomplete, also dirty as some people might not track all of their runs. A lot of confounders which might not have been taken into account due to the experiements (even with instrumental variabels). Biased as they only track people who aim to do sport which is not representative.

Nicolaides's pros: Large data set easier to find significant differences. 


#### 2

centola: the result doesn't nessicarily apply to different information subjects as peoples behavour differs alot when people for example are exposed to clickbait. They also can't really justify that its how it works in social networks as the influence friends have on each other is way different then random people, and especially when the random people is selected from a group who all are interested in the thing they are exposed to.  

Nicolaides's: Can lead to false conclussion as their might still be hidden confounders due to the lack of control as in custom-made data. However, as Nicolaides data is build on data which is not constructed in an arificial enviorment it is more correct to draw conclussions about the social structures within the world. 

# Part 3 Gathering Research Articles using the OpenAlex API

In [33]:
# Load the dataframe created in week 2 
df_authors = pd.read_csv('authors.csv')
display(df_authors)

id         display_name  \
0     https://openalex.org/A5037506201           Linda Steg   
1     https://openalex.org/A5027036879          Sharad Goel   
2     https://openalex.org/A5033639139          Tim Althoff   
3     https://openalex.org/A5005475843   Lauren J. N. Brent   
4     https://openalex.org/A5088539840         Esteban Moro   
...                                ...                  ...   
1224  https://openalex.org/A5032381602  Gabriella M. Harari   
1225  https://openalex.org/A5059015114      Coen D. Needell   
1226  https://openalex.org/A5084148692    Elliot E. Pickens   
1227  https://openalex.org/A5056216760          Anna George   
1228  https://openalex.org/A5004427266         Ana Valdivia   

                                          works_api_url  h_index  works_count  \
0     https://api.openalex.org/works?filter=author.i...       77          474   
1     https://api.openalex.org/works?filter=author.i...       37          176   
2     https://api.openalex.org/works?filter=author.i...       24          118   
3     https://api.openalex.org/works?filter=author.i...       27          122   
4     https://api.openalex.org/works?filter=author.i...       34          159   
...                                                 ...      ...          ...   
1224  https://api.openalex.org/works?filter=author.i...       20           78   
1225  https://api.openalex.org/works?filter=author.i...        2            3   
1226  https://api.openalex.org/works?filter=author.i...        1            2   
1227  https://api.openalex.org/works?filter=author.i...       29          149   
1228  https://api.openalex.org/works?filter=author.i...       10           22   

     country_code  
0              NL  
1              US  
2              US  
3              GB  
4              ES  
...           ...  
1224           US  
1225           US  
1226           GB  
1227           US  
1228           GB  

[1229 rows x 6 columns]

In [34]:
# we get the id's so we can filter using these

concepts = ['Sociology', 'Psychology', 'Economics', 'Political science', 'Mathematics', 'Physics', 'Computer science']
url = "https://api.openalex.org/concepts"
concept_id = []
for concept in concepts:
    response = requests.get(url, params={'search': concept, 'filter': 'level:0'})
    data = response.json()
    concept_id.append(data['results'][0]['id'])

[{'id': 'https://openalex.org/C144024400', 'wikidata': 'https://www.wikidata.org/wiki/Q21201', 'display_name': 'Sociology', 'relevance_score': 332977.7, 'level': 0, 'description': 'scientific study of human society and its origins, development, organizations, and institutions', 'works_count': 18361976, 'cited_by_count': 144951416, 'summary_stats': {'2yr_mean_citedness': 1.040750541760604, 'h_index': 2000, 'i10_index': 2356769}, 'ids': {'openalex': 'https://openalex.org/C144024400', 'wikidata': 'https://www.wikidata.org/wiki/Q21201', 'mag': '144024400', 'wikipedia': 'https://en.wikipedia.org/wiki/Sociology', 'umls_cui': ['C0037468']}, 'image_url': 'https://upload.wikimedia.org/wikipedia/commons/e/e4/Social_Network_Diagram_%28segment%29.svg', 'image_thumbnail_url': 'https://upload.wikimedia.org/wikipedia/commons/thumb/e/e4/Social_Network_Diagram_%28segment%29.svg/100px-Social_Network_Diagram_%28segment%29.svg.png', 'international': {'display_name': {'af': 'sosiologie', 'am': 'የባሕል ጥናት', 

In [35]:
df_authors = df_authors[(df_authors['works_count'] > 5) & (df_authors['works_count'] < 5000)]

# Request function for parallel processing
def request_works(author_ids: list[str]) -> tuple[pd.DataFrame, pd.DataFrame]:
    """
    Request works from the OpenAlex API for a list of authors.
    
    Args:
    author_names: list of author names (str)

    Returns:
    df_papers: DataFrame with the works of the authors
    df_abstracts: DataFrame with the abstracts of the works
    """
    # Initialize empty DataFrames for storing the data
    df_papers = pd.DataFrame()
    df_abstracts = pd.DataFrame()

    # For 20 authors at a time, make a request to the OpenAlex API to fetch their works
    author_id = "|".join(author_ids)
    url = "https://api.openalex.org/works?"
    params = {'per-page': 200, 
        'cursor': '*',
        'filter': f'authorships.author.id:{author_id}' + ',' + 
        'cited_by_count:>10' + ',' + 
        'authors_count:<10' + ',' +
        f'concept.id:{concept_ids[0]}|{concept_ids[1]}|{concept_ids[2]}|{concept_ids[3]}' + ',' +
        f'concept.id:{concept_ids[4]}|{concept_ids[5]}|{concept_ids[6]}' 
        }
    response = requests.get(url, params=params)
    data = response.json()

    # For each work, extract the relevant information and add it to the DataFrames, while going through the pages
    while data['meta']['next_cursor'] != None:
        if data['meta']['next_cursor'] == 0:
            break
        else:
            params["cursor"] = data['meta']["next_cursor"]
            for work in data['results']:
                
                df_papers_current = pd.DataFrame({
                    'id': [work['id']],
                    'publication_year': [work['publication_year']],
                    'cited_by_count': [work['cited_by_count']],
                    'author_ids': [[author['author']['id'] for author in work['authorships']]]
                })
                df_papers = pd.concat([df_papers, df_papers_current], ignore_index=True)

                df_abstracts_current = pd.DataFrame({
                    'id': [work['id']],
                    'title': [work['title']],
                    'abstract_inverted_index': [work['abstract_inverted_index']]
                })
                df_abstracts = pd.concat([df_abstracts, df_abstracts_current], ignore_index=True)

            response = requests.get(url, params=params)
            data = response.json()
    
    return df_papers, df_abstracts

# Parallel processing
results = Parallel(n_jobs=4)(delayed(request_works)(df_authors['id'][i:i+20]) for i in tqdm(range(0, len(df_authors['id']), 20), desc='Requesting data: '))
  

Requesting data: 100%|██████████| 51/51 [00:29<00:00,  1.72it/s]


In [37]:
df_papers = pd.concat([result[0] for result in results], ignore_index=True)
df_abstracts = pd.concat([result[1] for result in results], ignore_index=True)

print(df_papers.shape)
print(df_abstracts.shape)

# Save the DataFrames to CSV files
df_papers.to_csv('papers.csv', index=False)
df_abstracts.to_csv('abstracts.csv', index=False)

(8769, 4)
(8769, 3)


In [40]:
df_authors = pd.read_csv('author_ids.csv')
df_authors = df_authors.rename(columns={'id': 'author_id'})

# For every author id in the author ids column of df_papers, give the author id a row in the dataframe
df_expanded_papers = df_papers.explode('author_ids').rename(columns={'author_ids': 'author_id'})

# Unique author ids
unique_author_ids = df_expanded_papers['author_id'].unique()
print('Unique autor IDs: ',len(unique_author_ids))

# Load
df_coauthors = pd.read_csv('co-authors.csv')
df_coauthors = df_coauthors.rename(columns={'id': 'author_id'})

# Append df_authors to df_coauthors
df_authors_combined = pd.concat([df_coauthors, df_authors])

# Remove duplicate author ids
df_authors_combined = df_authors_combined.drop_duplicates(subset='author_id')

# Merge the two dataframes
df_merged = pd.merge(df_expanded_papers, df_authors_combined, on='author_id')

Unique autor IDs:  12476


Data Overview and Reflection questions: Answer the following questions:

Dataset summary. How many works are listed in your IC2S2 papers dataframe? How many unique researchers have co-authored these works?

- we have found 8769 papers, and a total of 12476

Efficiency in code. Describe the strategies you implemented to make your code more efficient. How did your approach affect your code's execution time? (answer in max 150 words)

- first of all, we created removed authors whose work count was outside the so we dont pull the data we dont need as the request takes the most time. We created all of the different filters to only extract the relevant data as we go through all the pages (which is a request call each time) this made the number of pages shrink = fewer API calls. We also take 200 per page rather than 25 as this again reduces API calls. lastly, we implemented the bulk request so we didn't take multiple researchers' works at once which again is more efficient because we overall have fewer pages as the pages more often are at 200. also added parallel to run 4 jobs at the same time.

Filtering Criteria and Dataset Relevance Reflect on the rationale behind setting specific thresholds for the total number of works by an author, the citation count, the number of authors per work, and the relevance of works to specific fields. How do these filtering criteria contribute to the relevance of the dataset you compiled? Do you believe any aspects of Computational Social Science research might be underrepresented or overrepresented as a result of these choices? (answer in max 150 words)

- firstly the threshold for works per author is good as we dont want someone who has barely touched the subject or once who is just flooding the internet with papers as this alludes to a low quality. Citation count is relevant as it is a great relevance parameter (if they never have been cited they are most likely not relevant). Also if there are more than 10 authors on a paper the contribution is likely low so might not be relevant to account for the different authors. It is also good to have the concept filter as we otherwise might get irrelevant data. 

- Maybe more niche social studies are excluded because the concept might be very specific and might not have the combination of concepts we check for (especially because we search for IDs so exact matches). 

# Part 4: The Network of Computational Social Scientists

In [45]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import networkx as nx
import os
from collections import defaultdict
import ast
import json

Getting the dict

In [55]:
# Load the papers.csv file
df_papers = pd.read_csv('papers.csv')
df_papers['author_ids'] = df_papers['author_ids'].apply(ast.literal_eval)

# Construct the adjacency list of the graph so we dont have to initialize when a new pair is found. 
adj_list = defaultdict(int)

for i in range(len(df_papers)):
    authors = df_papers['author_ids'][i]
    for j in range(len(authors)):
        for k in range(j+1, len(authors)):
            adj_list[(authors[j], authors[k])] += 1

Creating the graph

In [56]:
# Construct the graph
G = nx.Graph()
weighted_edge_list = [(u, v, d) for (u, v), d in adj_list.items()]
G.add_weighted_edges_from(weighted_edge_list)

Check if we still authors is loaded

In [57]:
# Add information to the graph nodes
for i in range(len(df_merged)):
    author_id = df_merged['author_id'][i]
    # Add the information to the graph if the node exists
    if author_id in G.nodes:
        G.nodes[author_id]['display_name'] = df_merged['display_name'][i]
        G.nodes[author_id]['country_code'] = df_merged['country_code'][i]
        G.nodes[author_id]['cited_by_count'] = int(df_merged['cited_by_count'][i])
        G.nodes[author_id]['publication_year'] = int(df_merged['publication_year'][i])

# Save the graph as json
graph_data = nx.node_link_data(G)
json.dump(graph_data, open('graph.json', 'w'))

1. Network Metrics:

In [58]:
# Total number of nodes (authors)
num_nodes = G.number_of_nodes()

# Total number of links (collaborations)
num_edges = G.number_of_edges()

# Network's density
density = nx.density(G)

# Check if the network is fully connected
is_connected = nx.is_connected(G)

# Number of connected components
num_connected_components = nx.number_connected_components(G)

# Number of isolated nodes
num_isolated_nodes = nx.number_of_isolates(G)

print(f"Number of nodes: {num_nodes}")
print(f"Number of edges: {num_edges}")
print(f"Density: {density}")
print(f"Is the network fully connected? {'Yes' if is_connected else 'No'}")
print(f"Number of connected components: {num_connected_components}")
print(f"Number of isolated nodes: {num_isolated_nodes}")

Number of nodes: 12467
Number of edges: 39305
Density: 0.0005058115176030065
Is the network fully connected? No
Number of connected components: 203
Number of isolated nodes: 0


Discuss the results above on network density, and connectivity. Are your findings in line with what you expected? Why?

- Based on the density we can conclude the network is extremely sparse. We had anticipated the network to be fully connected based on the number of collaborators in each work, however, the results make sense due to the concepts vary a lot and some might only focus on very niche subgroups under each concept. Also, the geographical distance makes sense for the network not being fully connected. However, we find that there are a lot of different connected components but this might also be due to the explanations above. It makes a lot of sense that no node is isolated because people often collaborate and we only focus on authors who have multiple works.

2. Degree Analysis:

In [59]:
# Get the degrees of the nodes
degrees = [deg for node, deg in G.degree()]

# Compute the statistics for degree
avg_degree = np.mean(degrees)
median_degree = np.median(degrees)
degree_uniq, degree_count = np.unique(degrees, return_counts=True)
mode_degree = degree_uniq[np.argmax(degree_count)]
min_degree = np.min(degrees)
max_degree = np.max(degrees)

# Get the weighted degrees (node strengths) of the nodes
strengths = [strength for node, strength in G.degree(weight='weight')]

# Compute the statistics for node strength
avg_strength = np.mean(strengths)
median_strength = np.median(strengths)
strength_uniq, strength_count = np.unique(strengths, return_counts=True)
mode_strength = strength_uniq[np.argmax(strength_count)]
min_strength = np.min(strengths)
max_strength = np.max(strengths)

print(f"Degree: Avg={avg_degree}, Median={median_degree}, Mode={mode_degree}, Min={min_degree}, Max={max_degree}")
print(f"Node strength: Avg={avg_strength}, Median={median_strength}, Mode={mode_strength}, Min={min_strength}, Max={max_strength}")

Degree: Avg=6.305446378439079, Median=5.0, Mode=3, Min=1, Max=169
Node strength: Avg=9.00601588192829, Median=5.0, Mode=3, Min=1, Max=463


Compute the average, median, mode, minimum, and maximum degree of the nodes. Perform the same analysis for node strength (weighted degree). What do these metrics tell us about the network? 

- That most researchers only collaborate with a certain handful of researchers. The average strength and average degree point to that they usually only collaborate twice with half of their collaboration, this could be due to research groups within the researchers. The most probable scenario is that you have 3 research partners you collaborate with once based on the mode. We also see that the max and min differ a lot and especially the max values are outliers due to the other numbers shown. 

3. Top Authors:

In [61]:
# Get the degrees of the nodes
degrees = G.degree()

# Sort the nodes by degree in descending order and get the top 5
top_authors = sorted(degrees, key=lambda x: x[1], reverse=True)[:5]

print("Top 5 authors by degree:")
for author, degree in top_authors:
    # Get the display name of the author
    display_name = G.nodes[author]['display_name']
    print(f"Author: {display_name}, Degree: {degree}, Author ID: {author}")

Top 5 authors by degree:
Author: Qin Li, Degree: 169, Author ID: https://openalex.org/A5075080019
Author: Mariano Sigman, Degree: 155, Author ID: https://openalex.org/A5005493160
Author: Stephan Lewandowsky, Degree: 150, Author ID: https://openalex.org/A5017914184
Author: Denny Borsboom, Degree: 140, Author ID: https://openalex.org/A5029100305
Author: Dan Jurafsky, Degree: 129, Author ID: https://openalex.org/A5087088138


Research these authors online. What areas do they specialize in? Do you think that their work aligns with the themes of Computational Social Science? If not, what could be possible reasons? 

- Qin Li - neurology/medicine - aligns fairly well she also has approximately 6 times as many works as the other so she might help out a lot of research and might be a part of a lot of social science but also research in other domains. 
- Mariano Sigman - neurology - aligns well with concepts due to the experiments and social aspect
- Stephan Lewandowsky - Political science with a focus on information flows. aligns with the quantitative and social science we have in the concepts 
- Denny Borsboom - psychological science with a focus on mental disease. Aligns with the concepts filtered for 
- Dan Jurafsky - Deep learning/machine learning - not social science it would be possible that there have been studies about interactions with deep learning / Machine learning / generative models. 

